src/pyFAI/gui/pilx/MainWindow.py

```diff
@@ -279,10 +282,25 @@ def displayImageAtIndices(self, indices: ImageIndices):
         if maskfile:
             mask_image = get_data(url=DataUrl(maskfile))
             if mask_image.shape != image.shape:
-                mask_image = None
+                binning = [m//i for i, m in zip(image.shape, mask_image.shape)]
+                if min(binning)<1:
+                    mask_image = None
+                else:
+                    mask_image = rebin_fct(mask_image, binning)
         else:
             mask_image = None
-
+        detector = self.worker_config.poni.detector
+        if detector:
+            detector_mask = detector.mask
+            if detector.shape != image.shape:
+                detector.guess_binning(image)
+                detector_mask = rebin_fct(detector_mask, detector.binning)
+            if mask_image is None:
+                mask_image = detector_mask
+            else:
+                numpy.logical_or(mask_image, detector_mask, out=mask_image)
+            detector.mask = mask_image
+            mask_image = detector.dynamic_mask(image)
```

Comment:
All these intertwined conditions make the code difficult to parse. I would suggest to move the highlighted part in a separate function (like `get_mask_file`) and use early returns. That will perhaps help a bit.